File: 14주차/DE/19. Remove Nth Node From End of List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
 function removeNthFromEnd(head, n) {
    if(head.next == null){
        return null
    }
    node = head
    cnt = 0
    while(node != null){
        node = node.next
        cnt += 1
    }
    node = head
    if(cnt - n - 1 < 0){
        return head = node.next
    }
    for(let i = 0; i < cnt - n - 1; i++){
        node = node.next
    }
    node.next = node.next.next
    return head;
};

File: 14주차/KM/19-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = head;
        ListNode prev = null;
        int cnt = 0;
        while(temp!=null){
            cnt++;
            temp = temp.next;
        }
        int target = cnt - n;
        cnt = 0;
        temp = head;
        while(cnt != target){
            prev = temp;
            temp = temp.next;
            cnt++;
        }
        if(prev == null){
            if(temp.next != null){
                return temp.next;
            } else {
                return null;
            }
        } else {
            prev.next = temp.next;
        }
        
        return head;
    }
}